src/Ledger/Utxo.lagda

```diff
@@ -230,11 +234,76 @@ certRefund (dereg c _)      = ❴ CredentialDeposit c ❵
-  = (updateCertDeposits pp certs deposits ∪⁺ certDeposit cert pp) ∣ certRefund cert ᶜ
+data ValidCertDeposits (pp : PParams) (deps : Deposits) : List DCert → Set
+\end{code}
+\begin{code}[hide]
+  where
+  []         : ValidCertDeposits pp deps []
+  delegate   : ∀ {c del kh v certs}
+             → ValidCertDeposits pp (deps ∪⁺ ❴ CredentialDeposit c , v ❵) certs
+             → ValidCertDeposits pp deps (delegate c del kh v ∷ certs)
+  regpool    : ∀ {kh p certs}
+             → ValidCertDeposits pp (deps ∪⁺ ❴ PoolDeposit kh , pp .poolDeposit ❵) certs
+             → ValidCertDeposits pp deps (regpool kh p ∷ certs)
+  regdrep    : ∀ {c v a certs}
+             → ValidCertDeposits pp (deps ∪⁺ ❴ DRepDeposit c , v ❵) certs
+             → ValidCertDeposits pp deps (regdrep c v a ∷ certs)
+  dereg      : ∀ {c d certs}
+             → (CredentialDeposit c , d) ∈ deps
+             → ValidCertDeposits pp (deps ∣ ❴ CredentialDeposit c ❵ ᶜ) certs
+             → ValidCertDeposits pp deps (dereg c d ∷ certs)
+  deregdrep  : ∀ {c d certs}
+             → (DRepDeposit c , d) ∈ deps
+             → ValidCertDeposits pp (deps ∣ ❴ DRepDeposit c ❵ ᶜ) certs
+             → ValidCertDeposits pp deps (deregdrep c d ∷ certs)
+  ccreghot   : ∀ {c v certs}
+             → ValidCertDeposits pp deps certs
+             → ValidCertDeposits pp deps (ccreghot c v ∷ certs)
+  retirepool : ∀ {kh e certs}
+             → ValidCertDeposits pp deps certs
+             → ValidCertDeposits pp deps (retirepool kh e  ∷ certs)
+
+private
+  validCertDeposits? : ∀ {pp} deps certs → Dec (ValidCertDeposits pp deps certs)
+  validCertDeposits? deps [] = yes []
+  validCertDeposits? deps (delegate _ _ _ _ ∷ certs) =
+    mapDec delegate (λ where (delegate p) → p) (validCertDeposits? _ _)
+  validCertDeposits? deps (regpool _ _ ∷ certs) =
+    mapDec regpool (λ where (regpool p) → p) (validCertDeposits? _ _)
+  validCertDeposits? deps (regdrep _ _ _ ∷ certs) =
+    mapDec regdrep (λ where (regdrep p) → p) (validCertDeposits? _ _)
+  validCertDeposits? deps (retirepool _ _ ∷ certs) =
+    mapDec retirepool (λ where (retirepool p) → p) (validCertDeposits? _ _)
+  validCertDeposits? deps (ccreghot _ _ ∷ certs) =
+    mapDec ccreghot (λ where (ccreghot p) → p) (validCertDeposits? _ _)
+  validCertDeposits? deps (dereg c d ∷ certs) with ¿ (CredentialDeposit c , d) ∈ deps ¿
+  ... | yes p = mapDec (dereg p)  (λ where (dereg _ d) → d) (validCertDeposits? _ _)
+  ... | no ¬p = no (λ where (dereg p _) → ¬p p)
+  validCertDeposits? deps (deregdrep c d ∷ certs) with ¿ (DRepDeposit c , d) ∈ deps ¿
+  ... | yes p = mapDec (deregdrep p)  (λ where (deregdrep _ v) → v) (validCertDeposits? _ _)
+  ... | no ¬p = no (λ where (deregdrep p _) → ¬p p)
+
+instance
+  Dec-ValidCertDeposits : ∀ {pp deps certs} → ValidCertDeposits pp deps certs ⁇
+  Dec-ValidCertDeposits = ⁇ (validCertDeposits? _ _)
+\end{code}
+\begin{code}
+
+-- Assumes ValidCertDeposits (mapˢ proj₁ (deposits ˢ)) certs.
```

Comment:
This comment should rather be explained in the prose.

Reply:
I've already added to the prose, "Suffice it to say that \ValidCertDeposits is used to check the validity of the deposits in a transaction so that the function \updateCertDeposits can correctly register and deregister deposits in the UTxO state based on the certificates in the transaction."  Perhaps that's enough.